### ISSUE TITLE
bug(THEME): Using css var() in palette contrast is not applied on all components

### ISSUE THREAD

**Issue by sschoorens** [angular/components#28716]
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### Description

When I defined a theme like that
```
@use '@angular/material' as mat;
@include mat.core();

:root {
  --primary: #2656cf;
  --primary-contrast: #FFFFFF;
}

$light-primary-palette: (
  700: var(--primary),
  500: var(--primary),
  100: var(--primary),
  contrast: (
    700: var(--primary-contrast),
    500: var(--primary-contrast),
    100: var(--primary-contrast)
  )
);

$light-theme: mat.define-light-theme((
 color: (
   primary: mat.define-palette($light-primary-palette),
   accent: mat.define-palette(mat.$pink-palette)
 ),
 typography: mat.define-typography-config( ),
 density: 0
));

@include mat.all-component-themes($light-theme);
```

Buttons don't apply the `--primary-contrast` but some other components use it

![image](https://github.com/angular/components/assets/728750/d843487d-4d4f-4618-8957-a3d2238a25c3)

Badges apply it
```
  --mat-badge-background-color: var(--primary);
  --mat-badge-text-color: var(--primary-contrast);
```

Buttons don't apply it
```
  --mdc-protected-button-container-color: var(--primary);
  --mdc-protected-button-label-text-color: #000;
  --mat-protected-button-state-layer-color: #000;
  --mat-protected-button-ripple-color: rgba(0, 0, 0, 0.1);
```

`--mdc-protected-button-label-text-color` should be `--mdc-protected-button-label-text-color:  var(--primary-contrast);`

### Reproduction

StackBlitz link:  https://stackblitz.com/edit/components-issue-u6e7fj
Steps to reproduce:
1. apply theme in styles.scss
2. look at the result on a primary button

### Expected Behavior

Button contrast should work

### Actual Behavior

Button contrast doesn't work

### Environment

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.1702.1
@angular-devkit/build-angular      17.2.1
@angular-devkit/core               17.2.1
@angular-devkit/schematics         17.2.1
@angular/cdk                       17.2.1
@angular/cli                       17.2.1
@angular/material                  17.2.1
@angular/material-moment-adapter   17.2.1
@schematics/angular                17.2.1
ng-packagr                         17.2.0
rxjs                               7.8.1
typescript                         5.3.3
zone.js                            0.14.4

**Comment by sschoorens** [comment]
Hovering buttons in mat-calendar also got the problem
![image](https://github.com/angular/components/assets/728750/115bc3d3-2cfc-44bd-8156-60d2514eb6e2)

**Comment by sschoorens** [comment]
Checkboxes too
![image](https://github.com/angular/components/assets/728750/2554adfb-9a7a-42d1-88d3-3cb8d0f876ca)

slide toggle also doesn't apply the primary color
![image](https://github.com/angular/components/assets/728750/fc69cd13-8a1c-4157-9ef9-2cca35783289)

**Comment by crisbeto** [comment]
Palettes based on CSS variables were never supported, see https://github.com/angular/components/issues/25981#issuecomment-1325334683. Instead of creating a palette with CSS variables that are initialized to colors, you can pass those colors directly to the palette and it'll create variables for you.

**Comment by crisbeto** [comment]
For what it's worth, testing it with the latest version (17.2.2) seems to mostly work, aside from the ripple colors which are also reported in https://github.com/angular/components/issues/28706.

**Comment by sschoorens** [comment]
Hello, thank you very much for all your answers and the fix in 17.2.2 also thanks for your fast work on #28706 that I also encounter.
Very good job !
